### PR TITLE
DT-270: Store and return user id

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/ResourceServerConfiguration.java [hunk 1]
```diff
@@ -83,6 +84,9 @@ public TokenStore tokenStore() {
     public JwtAccessTokenConverter accessTokenConverter() {
         final var converter = new JwtAccessTokenConverter();
         converter.setVerifierKey(new String(Base64.decodeBase64(properties.getJwtPublicKey())));
+        final var tokenConverter = new DefaultAccessTokenConverter();
+        tokenConverter.setUserTokenConverter(new UserIdAuthenticationConverter());
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/ResourceServerConfiguration.java]: need a new converter to grab the `user_id` out of the jwt

File: src/main/java/uk/gov/justice/hmpps/casenotes/services/CaseNoteService.java [hunk 2]
```diff
@@ -167,6 +169,7 @@ private CaseNote mapper(final NomisCaseNote cn, final String offenderIdentifier)
                 .occurrenceDateTime(cn.getOccurrenceDateTime())
                 .authorName(cn.getAuthorName())
                 .authorUsername(String.valueOf(cn.getStaffId()))  // TODO: this should be username not id.
+                .authorUserId(String.valueOf(cn.getStaffId()))
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/services/CaseNoteService.java]: I'll remove `authorUsername` in prisonstaffhub before removing it from here

File: src/main/resources/db/migration/general/V4_4__offender_case_note_user_id.sql [hunk 3]
```diff
@@ -0,0 +1,21 @@
+ALTER TABLE offender_case_note
+    ADD author_user_id VARCHAR(36) NULL;
+
+UPDATE offender_case_note
+SET author_user_id = author_username;
```
Comment (petergphillips) [inline src/main/resources/db/migration/general/V4_4__offender_case_note_user_id.sql]: note setting the `user_id` to be `username` for existing rows

Comment (Mjwillis) [reply src/main/resources/db/migration/general/V4_4__offender_case_note_user_id.sql]: 👍

File: src/main/java/uk/gov/justice/hmpps/casenotes/services/CaseNoteService.java (hunk 2 above)
Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/casenotes/services/CaseNoteService.java]: Ok

File: src/main/java/uk/gov/justice/hmpps/casenotes/config/SecurityUserContext.java [hunk 4]
```diff
@@ -32,11 +27,23 @@ private Authentication getAuthentication() {
...
+    public UserIdUser getCurrentUser() {
+        return getOptionalCurrentUser().orElseThrow(() -> new IllegalStateException("Current user not set but is required"));
+    }
 
-        final var userPrincipal = getUserPrincipal();
+    private Optional<UserIdUser> getOptionalCurrentUser() {
+        final var authentication = getAuthentication();
+        if (authentication == null || authentication.getPrincipal() == null) return Optional.empty();
 
+        final var userPrincipal = authentication.getPrincipal();
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/casenotes/config/SecurityUserContext.java]: Very very minor comment - might be cleaner to move this line ^^ above the null check and replace authentication.getPrincipal() == null with userPrincipal == null to be more consistent.

**Pull request merged**